💥 [DevOps] Fix `ocelotsocialnetwork/webapp:latest` And `ocelotsocialnetwork/backend:latest` On Start In Cluster

## 💥 DevOps Ticket
<!-- Describe your issue in detail. Include screenshots if needed. Give us as much information as possible. Use a clear and concise description of what the problem is.-->

Fix `ocelotsocialnetwork/webapp:latest` and `ocelotsocialnetwork/backend:latest` on start in cluster, see https://github.com/Ocelot-Social-Community/Ocelot-Social/issues/5065#issuecomment-1179732798

relates #5065

### Additiona Information

#### Dig For Image Informations And File Structure

Install Dive for Docker: https://github.com/wagoodman/dive

```bash
$ dive <image ID or image name>
```

#### Errors

Docker image `ocelotsocialnetwork/webapp:latest` on start in cluster errors with:

```bash
yarn run v1.22.5
$ nuxt start

 FATAL  Cannot find module './constants/manifest.js'
Require stack:
- /app/nuxt.config.js
- /app/node_modules/@nuxt/config/dist/config.js
- /app/node_modules/@nuxt/cli/dist/cli-index.js
- /app/node_modules/@nuxt/cli/dist/cli.js
- /app/node_modules/nuxt/bin/nuxt.js

  Error: Cannot find module './constants/manifest.js'
  Require stack:
  - nuxt.config.js
  - node_modules/@nuxt/config/dist/config.js
  - node_modules/@nuxt/cli/dist/cli-index.js
  - node_modules/@nuxt/cli/dist/cli.js
  - node_modules/nuxt/bin/nuxt.js
  at Object.<anonymous> (nuxt.config.js:1)

7[r[999;999H[6n8
   ╭─────────────────────────────────────────────────────────╮
   │                                                         │
   │   ✖ Nuxt Fatal Error                                    │
   │                                                         │
   │   Error: Cannot find module './constants/manifest.js'   │
   │   Require stack:                                        │
   │   - /app/nuxt.config.js                                 │
   │   - /app/node_modules/@nuxt/config/dist/config.js       │
   │   - /app/node_modules/@nuxt/cli/dist/cli-index.js       │
   │   - /app/node_modules/@nuxt/cli/dist/cli.js             │
   │   - /app/node_modules/nuxt/bin/nuxt.js                  │
   │                                                         │
   ╰─────────────────────────────────────────────────────────╯

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Docker image `ocelotsocialnetwork/backend:latest` on start in cluster errors with:

```bash
yarn run v1.22.5
$ node dist/
internal/modules/cjs/loader.js:834
  throw err;
  ^

Error: Cannot find module 'express'
Require stack:
- /app/dist/server.js
- /app/dist/index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:831:15)
    at Function.Module._load (internal/modules/cjs/loader.js:687:27)
    at Module.require (internal/modules/cjs/loader.js:903:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/app/dist/server.js:8:39)
    at Module._compile (internal/modules/cjs/loader.js:1015:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1035:10)
    at Module.load (internal/modules/cjs/loader.js:879:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:903:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/app/dist/server.js', '/app/dist/index.js' ]
}
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

--- a/DOCKER_MORE_CLOSELY.md
+++ b/DOCKER_MORE_CLOSELY.md
@@ -1,0 +1,64 @@
+# Docker More Closely
+
+## Apple M1 Platform
+
+***Attention:** For using Docker commands in Apple M1 environments!*
+
+### Enviroment Variable For Apple M1 Platform
+
+To set the Docker platform environment variable in your terminal tab, run:
+
+```bash
+# set env variable for your shell
+$ export DOCKER_DEFAULT_PLATFORM=linux/amd64
+```
+
+## Analysing Docker Builds
+
+To analyze a Docker build, there is a wonderful tool called [dive](https://github.com/wagoodman/dive). Please sponsor if you're using it!
+
+The `dive build` command is exactly the right one to fulfill what we are looking for.
+We can use it just like the `docker build` command and get an analysis afterwards.
+
+So, in our main folder, we use it in the following way:
+
+```bash
+# in main folder
+$ dive build --target <layer-name> -t "ocelotsocialnetwork/<app-name>:local-<layer-name>" --build-arg BBUILD_DATE="<build-date>" --build-arg BBUILD_VERSION="<build-version>" --build-arg BBUILD_COMMIT="<build-commit>" <app-folder-name-or-dot>/
+```
+
+The build arguments are optional.
+
+For the specific applications, we use them as follows.
+
+### Backend
+
+#### Production For Backend
+
+```bash
+# in main folder
+$ dive build --target production -t "ocelotsocialnetwork/backend:local-production" backend/
+```
+
+#### Development For Backend
+
+```bash
+# in main folder
+$ dive build --target development -t "ocelotsocialnetwork/backend:local-development" backend/
+```
+
+### Webapp
+
+#### Production For Webapp
+
+```bash
+# in main folder
+$ dive build --target production -t "ocelotsocialnetwork/webapp:local-production" webapp/
+```
+
+#### Development For Webapp
+
+```bash
+# in main folder
+$ dive build --target development -t "ocelotsocialnetwork/webapp:local-development" webapp/
+```

--- a/SUMMARY.md
+++ b/SUMMARY.md
@@ -15,6 +15,7 @@
   * [End-to-end tests](cypress/README.md)
   * [Frontend tests](webapp/testing.md)
   * [Backend tests](backend/testing.md)
+* [Docker More Closely](DOCKER_MORE_CLOSELY.md)
 * [Deployment](https://github.com/Ocelot-Social-Community/Ocelot-Social-Deploy-Rebranding/blob/master/deployment/README.md)
 * [Contributing](CONTRIBUTING.md)
 * [Feature Specification](cypress/features.md)

--- a/backend/Dockerfile
+++ b/backend/Dockerfile
@@ -91,6 +91,7 @@ FROM base as production
 
 # Copy "binary"-files from build image
 COPY --from=build ${DOCKER_WORKDIR}/dist ./dist
+COPY --from=build ${DOCKER_WORKDIR}/node_modules ./node_modules
 # Copy static files
 # TODO - externalize the uploads so we can copy the whole folder
 COPY --from=build ${DOCKER_WORKDIR}/public/img/ ./public/img/

--- a/webapp/Dockerfile
+++ b/webapp/Dockerfile
@@ -94,6 +94,13 @@ FROM base as production
 COPY --from=build ${DOCKER_WORKDIR}/.nuxt ./.nuxt
 COPY --from=build ${DOCKER_WORKDIR}/node_modules ./node_modules
 COPY --from=build ${DOCKER_WORKDIR}/nuxt.config.js ./nuxt.config.js
+# Copy static files
+# TODO - this seems not be needed anymore for the new rebranding
+# TODO - this should be one Folder containign all stuff needed to be copied
+COPY --from=build ${DOCKER_WORKDIR}/config/ ./config/
+COPY --from=build ${DOCKER_WORKDIR}/constants ./constants
+COPY --from=build ${DOCKER_WORKDIR}/static ./static
+COPY --from=build ${DOCKER_WORKDIR}/locales ./locales
 # Copy package.json for script definitions (lock file should not be needed)
 COPY --from=build ${DOCKER_WORKDIR}/package.json ./package.json
 

